diff --git c/CMakeLists.txt w/CMakeLists.txt
index f8d62e06b7..d420b0ec91 100644
--- c/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -1401,4 +1401,7 @@ endforeach()
 # to have correct paths to libraries depending on the configuration
 OCCT_UPDATE_TARGET_FILE ()
 
+# Check for system Wayland compatibility and compile run-time flag
+OCCT_INCLUDE_CMAKE_FILE("adm/cmake/wayland_compat")
+
 message (STATUS "Info: \(${CURRENT_TIME}\) OCCT configuration files prepared")
diff --git c/adm/cmake/wayland_compat.cmake w/adm/cmake/wayland_compat.cmake
new file mode 100644
index 0000000000..99e51710e5
--- /dev/null
+++ w/adm/cmake/wayland_compat.cmake
@@ -0,0 +1,15 @@
+# Compile definition for Wayland compatibility
+if(UNIX
+   AND NOT USE_GLES2
+   AND NOT APPLE
+   AND NOT ANDROID)
+  set(WAYLAND_COMPAT ON)
+  foreach(target ${OCCT_LIBRARIES})
+    target_compile_definitions(${target} PRIVATE WAYLAND_COMPAT=1)
+  endforeach()
+else()
+  set(WAYLAND_COMPAT OFF)
+  foreach(target ${OCCT_LIBRARIES})
+    target_compile_definitions(${target} PRIVATE WAYLAND_COMPAT=0)
+  endforeach()
+endif()
diff --git c/src/Aspect/Aspect_DisplayConnection.cxx w/src/Aspect/Aspect_DisplayConnection.cxx
index ce6ee9ac4f..c58ec1d3d4 100755
--- c/src/Aspect/Aspect_DisplayConnection.cxx
+++ w/src/Aspect/Aspect_DisplayConnection.cxx
@@ -15,8 +15,9 @@
 
 #include <Aspect_DisplayConnectionDefinitionError.hxx>
 #include <OSD_Environment.hxx>
+#include <Standard_Macro_Runtime.hxx>
 
-#if defined(HAVE_XLIB)
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
   #include <X11/Xlib.h>
   #include <X11/Xutil.h>
 #endif
@@ -27,14 +28,17 @@ IMPLEMENT_STANDARD_RTTIEXT(Aspect_DisplayConnection, Standard_Transient)
 
 Aspect_DisplayConnection::Aspect_DisplayConnection()
 {
-#if defined(HAVE_XLIB)
-  myDisplay       = NULL;
-  myDefVisualInfo = NULL;
-  myDefFBConfig   = NULL;
-  myIsOwnDisplay  = false;
-  OSD_Environment anEnv("DISPLAY");
-  myDisplayName = anEnv.Value();
-  Init(NULL);
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (!RUN_WAYLAND)
+  {
+    myDisplay       = NULL;
+    myDefVisualInfo = NULL;
+    myDefFBConfig   = NULL;
+    myIsOwnDisplay  = false;
+    OSD_Environment anEnv("DISPLAY");
+    myDisplayName = anEnv.Value();
+    Init(NULL);
+  }
 #endif
 }
 
@@ -42,14 +46,17 @@ Aspect_DisplayConnection::Aspect_DisplayConnection()
 
 Aspect_DisplayConnection::~Aspect_DisplayConnection()
 {
-#if defined(HAVE_XLIB)
-  if (myDefVisualInfo != NULL)
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (!RUN_WAYLAND)
   {
-    XFree(myDefVisualInfo);
-  }
-  if (myDisplay != NULL && myIsOwnDisplay)
-  {
-    XCloseDisplay((Display*)myDisplay);
+    if (myDefVisualInfo != NULL)
+    {
+      XFree(myDefVisualInfo);
+    }
+    if (myDisplay != NULL && myIsOwnDisplay)
+    {
+      XCloseDisplay((Display*)myDisplay);
+    }
   }
 #endif
 }
@@ -84,8 +91,9 @@ void Aspect_DisplayConnection::SetDefaultVisualInfo(Aspect_XVisualInfo* theVisua
 {
   if (myDefVisualInfo != NULL)
   {
-#if defined(HAVE_XLIB)
-    XFree(myDefVisualInfo);
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+    if (!RUN_WAYLAND)
+      XFree(myDefVisualInfo);
 #endif
   }
   myDefVisualInfo = theVisual;
@@ -96,31 +104,38 @@ void Aspect_DisplayConnection::SetDefaultVisualInfo(Aspect_XVisualInfo* theVisua
 
 void Aspect_DisplayConnection::Init(Aspect_XDisplay* theDisplay)
 {
-#if defined(HAVE_XLIB)
-  if (myDisplay != NULL && myIsOwnDisplay)
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (!RUN_WAYLAND)
   {
-    XCloseDisplay((Display*)myDisplay);
-  }
-  myIsOwnDisplay = false;
-  myAtoms.Clear();
+    if (myDisplay != NULL && myIsOwnDisplay)
+    {
+      XCloseDisplay((Display*)myDisplay);
+    }
+    myIsOwnDisplay = false;
+    myAtoms.Clear();
 
-  myDisplay =
-    theDisplay != NULL ? theDisplay : (Aspect_XDisplay*)XOpenDisplay(myDisplayName.ToCString());
-  if (myDisplay == NULL)
-  {
-    TCollection_AsciiString aMessage;
-    aMessage += "Can not connect to the server \"";
-    aMessage += myDisplayName + "\"";
-    throw Aspect_DisplayConnectionDefinitionError(aMessage.ToCString());
+    myDisplay =
+      theDisplay != NULL ? theDisplay : (Aspect_XDisplay*)XOpenDisplay(myDisplayName.ToCString());
+    if (myDisplay == NULL)
+    {
+      TCollection_AsciiString aMessage;
+      aMessage += "Can not connect to the server \"";
+      aMessage += myDisplayName + "\"";
+      throw Aspect_DisplayConnectionDefinitionError(aMessage.ToCString());
+    }
+    else
+    {
+      myIsOwnDisplay = theDisplay == NULL;
+      myAtoms.Bind(Aspect_XA_DELETE_WINDOW,
+                   (uint64_t)XInternAtom((Display*)myDisplay, "WM_DELETE_WINDOW", False));
+    }
   }
-  else
+#endif
+#if !defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (!RUN_UNIX_X11)
   {
+    myDisplay      = theDisplay;
     myIsOwnDisplay = theDisplay == NULL;
-    myAtoms.Bind(Aspect_XA_DELETE_WINDOW,
-                 (uint64_t)XInternAtom((Display*)myDisplay, "WM_DELETE_WINDOW", False));
   }
-#else
-  myDisplay      = theDisplay;
-  myIsOwnDisplay = theDisplay == NULL;
 #endif
 }
diff --git c/src/OpenGl/OpenGl_Context.cxx w/src/OpenGl/OpenGl_Context.cxx
index 61c306076c..657c475e4c 100644
--- c/src/OpenGl/OpenGl_Context.cxx
+++ w/src/OpenGl/OpenGl_Context.cxx
@@ -53,7 +53,11 @@
 
 IMPLEMENT_STANDARD_RTTIEXT(OpenGl_Context, Standard_Transient)
 
-#if defined(HAVE_EGL)
+#include <Standard_Macro_Runtime.hxx>
+#if WAYLAND_COMPAT
+  #include <EGL/egl.h>
+  #include <GL/glx.h>
+#elif defined(HAVE_EGL)
   #include <EGL/egl.h>
   #ifdef _MSC_VER
     #pragma comment(lib, "libEGL.lib")
@@ -267,10 +271,13 @@ OpenGl_Context::OpenGl_Context(const Handle(OpenGl_Caps)& theCaps)
   myPolygonOffset.Units  = 0.0f;
 
   // system-dependent fields
-#if defined(HAVE_EGL)
-  myDisplay  = (Aspect_Display)EGL_NO_DISPLAY;
-  myWindow   = (Aspect_Drawable)EGL_NO_SURFACE;
-  myGContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
+#if defined(COMPILE_EGL)
+  if (!RUN_UNIX_X11)
+  {
+    myDisplay  = (Aspect_Display)EGL_NO_DISPLAY;
+    myWindow   = (Aspect_Drawable)EGL_NO_SURFACE;
+    myGContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
+  };
 #elif defined(__APPLE__) && !defined(HAVE_XLIB)
   // Vendors can not extend functionality on this system
   // and developers are limited to OpenGL support provided by Mac OS X SDK.
@@ -622,31 +629,38 @@ void OpenGl_Context::Share(const Handle(OpenGl_Context)& theShareCtx)
 
 Standard_Boolean OpenGl_Context::IsCurrent() const
 {
-  #if defined(HAVE_EGL)
-  if ((EGLDisplay)myDisplay == EGL_NO_DISPLAY || (EGLContext)myGContext == EGL_NO_CONTEXT)
+  #ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11)
   {
-    return Standard_False;
-  }
-
-  return (((EGLDisplay)myDisplay == eglGetCurrentDisplay())
-          && ((EGLContext)myGContext == eglGetCurrentContext())
-          && ((EGLSurface)myWindow == eglGetCurrentSurface(EGL_DRAW)));
-  #elif defined(_WIN32)
+    if ((EGLDisplay)myDisplay == EGL_NO_DISPLAY || (EGLContext)myGContext == EGL_NO_CONTEXT)
+    {
+      return Standard_False;
+    }
+    return (((EGLDisplay)myDisplay == eglGetCurrentDisplay())
+            && ((EGLContext)myGContext == eglGetCurrentContext())
+            && ((EGLSurface)myWindow == eglGetCurrentSurface(EGL_DRAW)));
+  };
+  #endif
+  #ifdef COMPILE_WIN32
   if (myDisplay == NULL || myGContext == NULL)
   {
     return Standard_False;
   }
   return (((HDC)myDisplay == wglGetCurrentDC()) && ((HGLRC)myGContext == wglGetCurrentContext()));
-  #elif defined(HAVE_XLIB)
-  if (myDisplay == NULL || myWindow == 0 || myGContext == 0)
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    return Standard_False;
-  }
-
-  return (((Display*)myDisplay == glXGetCurrentDisplay())
-          && ((GLXContext)myGContext == glXGetCurrentContext())
-          && ((GLXDrawable)myWindow == glXGetCurrentDrawable()));
-  #else
+    if (myDisplay == NULL || myWindow == 0 || myGContext == 0)
+    {
+      return Standard_False;
+    }
+    return (((Display*)myDisplay == glXGetCurrentDisplay())
+            && ((GLXContext)myGContext == glXGetCurrentContext())
+            && ((GLXDrawable)myWindow == glXGetCurrentDrawable()));
+  };
+  #endif
+  #ifdef COMPILE_NOOP
   return Standard_False;
   #endif
 }
@@ -655,30 +669,34 @@ Standard_Boolean OpenGl_Context::IsCurrent() const
 
 Standard_Boolean OpenGl_Context::MakeCurrent()
 {
-  #if defined(HAVE_EGL)
-  if ((EGLDisplay)myDisplay == EGL_NO_DISPLAY || (EGLContext)myGContext == EGL_NO_CONTEXT)
+  #ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11)
   {
-    Standard_ProgramError_Raise_if(myIsInitialized,
-                                   "OpenGl_Context::Init() should be called before!");
-    return Standard_False;
-  }
+    if ((EGLDisplay)myDisplay == EGL_NO_DISPLAY || (EGLContext)myGContext == EGL_NO_CONTEXT)
+    {
+      Standard_ProgramError_Raise_if(myIsInitialized,
+                                     "OpenGl_Context::Init() should be called before!");
+      return Standard_False;
+    }
 
-  if (eglMakeCurrent((EGLDisplay)myDisplay,
-                     (EGLSurface)myWindow,
-                     (EGLSurface)myWindow,
-                     (EGLContext)myGContext)
-      != EGL_TRUE)
-  {
-    // if there is no current context it might be impossible to use glGetError() correctly
-    PushMessage(GL_DEBUG_SOURCE_WINDOW_SYSTEM,
-                GL_DEBUG_TYPE_ERROR,
-                0,
-                GL_DEBUG_SEVERITY_HIGH,
-                "eglMakeCurrent() has failed!");
-    myIsInitialized = Standard_False;
-    return Standard_False;
+    if (eglMakeCurrent((EGLDisplay)myDisplay,
+                       (EGLSurface)myWindow,
+                       (EGLSurface)myWindow,
+                       (EGLContext)myGContext)
+        != EGL_TRUE)
+    {
+      // if there is no current context it might be impossible to use glGetError() correctly
+      PushMessage(GL_DEBUG_SOURCE_WINDOW_SYSTEM,
+                  GL_DEBUG_TYPE_ERROR,
+                  0,
+                  GL_DEBUG_SEVERITY_HIGH,
+                  "eglMakeCurrent() has failed!");
+      myIsInitialized = Standard_False;
+      return Standard_False;
+    }
   }
-  #elif defined(_WIN32)
+  #endif
+  #ifdef COMPILE_WIN32
   if (myDisplay == NULL || myGContext == NULL)
   {
     Standard_ProgramError_Raise_if(myIsInitialized,
@@ -720,26 +738,31 @@ Standard_Boolean OpenGl_Context::MakeCurrent()
     myIsInitialized = Standard_False;
     return Standard_False;
   }
-  #elif defined(HAVE_XLIB)
-  if (myDisplay == NULL || myWindow == 0 || myGContext == 0)
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    Standard_ProgramError_Raise_if(myIsInitialized,
-                                   "OpenGl_Context::Init() should be called before!");
-    return Standard_False;
-  }
+    if (myDisplay == NULL || myWindow == 0 || myGContext == 0)
+    {
+      Standard_ProgramError_Raise_if(myIsInitialized,
+                                     "OpenGl_Context::Init() should be called before!");
+      return Standard_False;
+    }
 
-  if (!glXMakeCurrent((Display*)myDisplay, (GLXDrawable)myWindow, (GLXContext)myGContext))
-  {
-    // if there is no current context it might be impossible to use glGetError() correctly
-    PushMessage(GL_DEBUG_SOURCE_WINDOW_SYSTEM,
-                GL_DEBUG_TYPE_ERROR,
-                0,
-                GL_DEBUG_SEVERITY_HIGH,
-                "glXMakeCurrent() has failed!");
-    myIsInitialized = Standard_False;
-    return Standard_False;
+    if (!glXMakeCurrent((Display*)myDisplay, (GLXDrawable)myWindow, (GLXContext)myGContext))
+    {
+      // if there is no current context it might be impossible to use glGetError() correctly
+      PushMessage(GL_DEBUG_SOURCE_WINDOW_SYSTEM,
+                  GL_DEBUG_TYPE_ERROR,
+                  0,
+                  GL_DEBUG_SEVERITY_HIGH,
+                  "glXMakeCurrent() has failed!");
+      myIsInitialized = Standard_False;
+      return Standard_False;
+    }
   }
-  #else
+  #endif
+  #ifdef COMPILE_NOOP
   // not implemented
   if (!myIsInitialized)
   {
@@ -754,23 +777,26 @@ Standard_Boolean OpenGl_Context::MakeCurrent()
 
 void OpenGl_Context::SwapBuffers()
 {
-  #if defined(HAVE_EGL)
-  if ((EGLSurface)myWindow != EGL_NO_SURFACE)
+  #ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11 && (EGLSurface)myWindow != EGL_NO_SURFACE)
   {
     eglSwapBuffers((EGLDisplay)myDisplay, (EGLSurface)myWindow);
   }
-  #elif defined(_WIN32)
+  #endif
+  #ifdef COMPILE_WIN32
   if ((HDC)myDisplay != NULL)
   {
     ::SwapBuffers((HDC)myDisplay);
     core11fwd->glFlush();
   }
-  #elif defined(HAVE_XLIB)
-  if ((Display*)myDisplay != NULL)
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND && (Display*)myDisplay != NULL)
   {
     glXSwapBuffers((Display*)myDisplay, (GLXDrawable)myWindow);
   }
-  #else
+  #endif
+  #ifdef COMPILE_NOOP
     //
   #endif
 }
@@ -781,39 +807,48 @@ void OpenGl_Context::SwapBuffers()
 
 Standard_Boolean OpenGl_Context::SetSwapInterval(const Standard_Integer theInterval)
 {
-#if defined(HAVE_EGL)
-  if (::eglSwapInterval((EGLDisplay)myDisplay, theInterval) == EGL_TRUE)
+#ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11 && ::eglSwapInterval((EGLDisplay)myDisplay, theInterval) == EGL_TRUE)
   {
     return Standard_True;
   }
-#elif defined(_WIN32)
+#endif
+#ifdef COMPILE_WIN32
   if (myFuncs->wglSwapIntervalEXT != NULL)
   {
     myFuncs->wglSwapIntervalEXT(theInterval);
     return Standard_True;
   }
-#elif defined(TARGET_OS_IPHONE) && TARGET_OS_IPHONE
+#endif
+#ifdef COMPILE_APPLE
+  #if defined(TARGET_OS_IPHONE) && TARGET_OS_IPHONE
   (void)theInterval; // vsync cannot be turned OFF on iOS
-#elif defined(__APPLE__)
+  #else
   if (::CGLSetParameter(CGLGetCurrentContext(), kCGLCPSwapInterval, &theInterval) == kCGLNoError)
   {
     return Standard_True;
   }
-#elif defined(HAVE_XLIB)
-  if (theInterval == -1 && myFuncs->glXSwapIntervalEXT != NULL)
+  #endif
+#endif
+#ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    typedef int (
-      *glXSwapIntervalEXT_t_x)(Display* theDisplay, GLXDrawable theDrawable, int theInterval);
-    glXSwapIntervalEXT_t_x aFuncPtr = (glXSwapIntervalEXT_t_x)myFuncs->glXSwapIntervalEXT;
-    aFuncPtr((Display*)myDisplay, (GLXDrawable)myWindow, theInterval);
-    return Standard_True;
+    if (theInterval == -1 && myFuncs->glXSwapIntervalEXT != NULL)
+    {
+      typedef int (
+        *glXSwapIntervalEXT_t_x)(Display* theDisplay, GLXDrawable theDrawable, int theInterval);
+      glXSwapIntervalEXT_t_x aFuncPtr = (glXSwapIntervalEXT_t_x)myFuncs->glXSwapIntervalEXT;
+      aFuncPtr((Display*)myDisplay, (GLXDrawable)myWindow, theInterval);
+      return Standard_True;
+    }
+    else if (myFuncs->glXSwapIntervalSGI != NULL)
+    {
+      myFuncs->glXSwapIntervalSGI(theInterval);
+      return Standard_True;
+    }
   }
-  else if (myFuncs->glXSwapIntervalSGI != NULL)
-  {
-    myFuncs->glXSwapIntervalSGI(theInterval);
-    return Standard_True;
-  }
-#else
+#endif
+#ifdef COMPILE_NOOP
   //
 #endif
   return Standard_False;
@@ -823,15 +858,21 @@ Standard_Boolean OpenGl_Context::SetSwapInterval(const Standard_Integer theInter
 
 void* OpenGl_Context::findProc(const char* theFuncName)
 {
-#if defined(HAVE_EGL)
-  return (void*)eglGetProcAddress(theFuncName);
-#elif defined(_WIN32)
+#ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11)
+    return (void*)eglGetProcAddress(theFuncName);
+#endif
+#ifdef COMPILE_WIN32
   return (void*)wglGetProcAddress(theFuncName);
-#elif defined(HAVE_XLIB)
-  return (void*)glXGetProcAddress((const GLubyte*)theFuncName);
-#elif defined(__APPLE__)
+#endif
+#ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
+    return (void*)glXGetProcAddress((const GLubyte*)theFuncName);
+#endif
+#ifdef COMPILE_APPLE
   return (myGlLibHandle != NULL) ? dlsym(myGlLibHandle, theFuncName) : NULL;
-#else
+#endif
+#ifdef COMPILE_NOOP
   (void)theFuncName;
   return NULL;
 #endif
@@ -940,18 +981,27 @@ Standard_Boolean OpenGl_Context::Init(const Standard_Boolean theIsCoreProfile)
     return Standard_True;
   }
 
-  #if defined(HAVE_EGL)
-  myDisplay  = (Aspect_Display)eglGetCurrentDisplay();
-  myGContext = (Aspect_RenderingContext)eglGetCurrentContext();
-  myWindow   = (Aspect_Drawable)eglGetCurrentSurface(EGL_DRAW);
-  #elif defined(_WIN32)
+  #ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11)
+  {
+    myDisplay  = (Aspect_Display)eglGetCurrentDisplay();
+    myGContext = (Aspect_RenderingContext)eglGetCurrentContext();
+    myWindow   = (Aspect_Drawable)eglGetCurrentSurface(EGL_DRAW);
+  }
+  #endif
+  #ifdef COMPILE_WIN32
   myDisplay  = (Aspect_Handle)wglGetCurrentDC();
   myGContext = (Aspect_RenderingContext)wglGetCurrentContext();
-  #elif defined(HAVE_XLIB)
-  myDisplay  = (Aspect_Display)glXGetCurrentDisplay();
-  myGContext = (Aspect_RenderingContext)glXGetCurrentContext();
-  myWindow   = (Aspect_Drawable)glXGetCurrentDrawable();
-  #else
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
+  {
+    myDisplay  = (Aspect_Display)glXGetCurrentDisplay();
+    myGContext = (Aspect_RenderingContext)glXGetCurrentContext();
+    myWindow   = (Aspect_Drawable)glXGetCurrentDrawable();
+  }
+  #endif
+  #ifdef COMPILE_NOOP
     //
   #endif
   if (myGContext == NULL)
@@ -1816,9 +1866,9 @@ void OpenGl_Context::MemoryInfo(TColStd_IndexedDataMapOfStringString& theDict) c
   }
 #endif
 
-#if defined(HAVE_XLIB) && !defined(__APPLE__) && !defined(_WIN32)
+#if defined(COMPILE_XLIB) && !defined(__APPLE__) && !defined(_WIN32)
   // GLX_RENDERER_VENDOR_ID_MESA
-  if (myFuncs->glXQueryCurrentRendererIntegerMESA != NULL)
+  if (!RUN_WAYLAND && myFuncs->glXQueryCurrentRendererIntegerMESA != NULL)
   {
     unsigned int aVMemMiB = 0;
     if (myFuncs->glXQueryCurrentRendererIntegerMESA(GLX_RENDERER_VIDEO_MEMORY_MESA, &aVMemMiB) != 0)
@@ -1846,9 +1896,10 @@ void OpenGl_Context::WindowBufferBits(Graphic3d_Vec4i& theColorBits,
   }
   else
   {
-#if defined(HAVE_EGL)
-    //
-#elif defined(_WIN32)
+#ifdef COMPILE_EGL
+//
+#endif
+#ifdef COMPILE_WIN32
     const int             aPixFrmtIndex = GetPixelFormat((HDC)myDisplay);
     PIXELFORMATDESCRIPTOR aFormat;
     memset(&aFormat, 0, sizeof(aFormat));
@@ -1859,25 +1910,29 @@ void OpenGl_Context::WindowBufferBits(Graphic3d_Vec4i& theColorBits,
                            aFormat.cBlueBits,
                            aFormat.cAlphaBits);
     theDepthStencilBits.SetValues(aFormat.cDepthBits, aFormat.cStencilBits);
-#elif defined(HAVE_XLIB)
-    Display*          aDisplay = (Display*)myDisplay;
-    XWindowAttributes aWinAttribs;
-    XGetWindowAttributes(aDisplay, (::Window)myWindow, &aWinAttribs);
-    XVisualInfo aVisInfo;
-    aVisInfo.visualid                                     = aWinAttribs.visual->visualid;
-    aVisInfo.screen                                       = DefaultScreen(aDisplay);
-    int                                          aNbItems = 0;
-    std::unique_ptr<XVisualInfo, int (*)(void*)> aVis(
-      XGetVisualInfo(aDisplay, VisualIDMask | VisualScreenMask, &aVisInfo, &aNbItems),
-      &XFree);
-    if (aVis.get() != NULL)
+#endif
+#ifdef COMPILE_XLIB
+    if (!RUN_WAYLAND)
     {
-      glXGetConfig(aDisplay, aVis.get(), GLX_RED_SIZE, &theColorBits.r());
-      glXGetConfig(aDisplay, aVis.get(), GLX_GREEN_SIZE, &theColorBits.g());
-      glXGetConfig(aDisplay, aVis.get(), GLX_BLUE_SIZE, &theColorBits.b());
-      glXGetConfig(aDisplay, aVis.get(), GLX_ALPHA_SIZE, &theColorBits.a());
-      glXGetConfig(aDisplay, aVis.get(), GLX_DEPTH_SIZE, &theDepthStencilBits[0]);
-      glXGetConfig(aDisplay, aVis.get(), GLX_STENCIL_SIZE, &theDepthStencilBits[1]);
+      Display*          aDisplay = (Display*)myDisplay;
+      XWindowAttributes aWinAttribs;
+      XGetWindowAttributes(aDisplay, (::Window)myWindow, &aWinAttribs);
+      XVisualInfo aVisInfo;
+      aVisInfo.visualid                                     = aWinAttribs.visual->visualid;
+      aVisInfo.screen                                       = DefaultScreen(aDisplay);
+      int                                          aNbItems = 0;
+      std::unique_ptr<XVisualInfo, int (*)(void*)> aVis(
+        XGetVisualInfo(aDisplay, VisualIDMask | VisualScreenMask, &aVisInfo, &aNbItems),
+        &XFree);
+      if (aVis.get() != NULL)
+      {
+        glXGetConfig(aDisplay, aVis.get(), GLX_RED_SIZE, &theColorBits.r());
+        glXGetConfig(aDisplay, aVis.get(), GLX_GREEN_SIZE, &theColorBits.g());
+        glXGetConfig(aDisplay, aVis.get(), GLX_BLUE_SIZE, &theColorBits.b());
+        glXGetConfig(aDisplay, aVis.get(), GLX_ALPHA_SIZE, &theColorBits.a());
+        glXGetConfig(aDisplay, aVis.get(), GLX_DEPTH_SIZE, &theDepthStencilBits[0]);
+        glXGetConfig(aDisplay, aVis.get(), GLX_STENCIL_SIZE, &theDepthStencilBits[1]);
+      }
     }
 #endif
   }
@@ -1890,42 +1945,51 @@ void OpenGl_Context::DiagnosticInformation(TColStd_IndexedDataMapOfStringString&
 {
   if ((theFlags & Graphic3d_DiagnosticInfo_NativePlatform) != 0)
   {
-#if defined(HAVE_EGL)
-    addInfo(theDict, "EGLVersion", ::eglQueryString((EGLDisplay)myDisplay, EGL_VERSION));
-    addInfo(theDict, "EGLVendor", ::eglQueryString((EGLDisplay)myDisplay, EGL_VENDOR));
-    addInfo(theDict, "EGLClientAPIs", ::eglQueryString((EGLDisplay)myDisplay, EGL_CLIENT_APIS));
-    if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
+#ifdef COMPILE_EGL
+    if (!RUN_UNIX_X11)
     {
-      addInfo(theDict, "EGLExtensions", ::eglQueryString((EGLDisplay)myDisplay, EGL_EXTENSIONS));
+      addInfo(theDict, "EGLVersion", ::eglQueryString((EGLDisplay)myDisplay, EGL_VERSION));
+      addInfo(theDict, "EGLVendor", ::eglQueryString((EGLDisplay)myDisplay, EGL_VENDOR));
+      addInfo(theDict, "EGLClientAPIs", ::eglQueryString((EGLDisplay)myDisplay, EGL_CLIENT_APIS));
+      if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
+      {
+        addInfo(theDict, "EGLExtensions", ::eglQueryString((EGLDisplay)myDisplay, EGL_EXTENSIONS));
+      }
     }
-#elif defined(_WIN32)
+#endif
+#ifdef COMPILE_WIN32
     if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0
         && myFuncs->wglGetExtensionsStringARB != NULL)
     {
       const char* aWglExts = myFuncs->wglGetExtensionsStringARB((HDC)myDisplay);
       addInfo(theDict, "WGLExtensions", aWglExts);
     }
-#elif defined(HAVE_XLIB)
-    Display*  aDisplay = (Display*)myDisplay;
-    const int aScreen  = DefaultScreen(aDisplay);
-    addInfo(theDict,
-            "GLXDirectRendering",
-            ::glXIsDirect(aDisplay, (GLXContext)myGContext) ? "Yes" : "No");
-    addInfo(theDict, "GLXVendor", ::glXQueryServerString(aDisplay, aScreen, GLX_VENDOR));
-    addInfo(theDict, "GLXVersion", ::glXQueryServerString(aDisplay, aScreen, GLX_VERSION));
-    if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
+#endif
+#ifdef COMPILE_XLIB
+    if (!RUN_WAYLAND)
     {
-      const char* aGlxExts = ::glXQueryExtensionsString(aDisplay, aScreen);
-      addInfo(theDict, "GLXExtensions", aGlxExts);
-    }
+      Display*  aDisplay = (Display*)myDisplay;
+      const int aScreen  = DefaultScreen(aDisplay);
+      addInfo(theDict,
+              "GLXDirectRendering",
+              ::glXIsDirect(aDisplay, (GLXContext)myGContext) ? "Yes" : "No");
+      addInfo(theDict, "GLXVendor", ::glXQueryServerString(aDisplay, aScreen, GLX_VENDOR));
+      addInfo(theDict, "GLXVersion", ::glXQueryServerString(aDisplay, aScreen, GLX_VERSION));
+      if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
+      {
+        const char* aGlxExts = ::glXQueryExtensionsString(aDisplay, aScreen);
+        addInfo(theDict, "GLXExtensions", aGlxExts);
+      }
 
-    addInfo(theDict, "GLXClientVendor", ::glXGetClientString(aDisplay, GLX_VENDOR));
-    addInfo(theDict, "GLXClientVersion", ::glXGetClientString(aDisplay, GLX_VERSION));
-    if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
-    {
-      addInfo(theDict, "GLXClientExtensions", ::glXGetClientString(aDisplay, GLX_EXTENSIONS));
+      addInfo(theDict, "GLXClientVendor", ::glXGetClientString(aDisplay, GLX_VENDOR));
+      addInfo(theDict, "GLXClientVersion", ::glXGetClientString(aDisplay, GLX_VERSION));
+      if ((theFlags & Graphic3d_DiagnosticInfo_Extensions) != 0)
+      {
+        addInfo(theDict, "GLXClientExtensions", ::glXGetClientString(aDisplay, GLX_EXTENSIONS));
+      }
     }
-#else
+#endif
+#ifdef COMPILE_NOOP
     //
 #endif
   }
@@ -2359,11 +2423,11 @@ void OpenGl_Context::SetShadingMaterial(
 
   // do not update material properties in case of zero reflection mode,
   // because GL lighting will be disabled by OpenGl_PrimitiveArray::DrawArray() anyway.
-  const OpenGl_MaterialState& aMatState     = myShaderManager->MaterialState();
-  float                       anAlphaCutoff = (anAspect->AlphaMode() == Graphic3d_AlphaMode_Mask
+  const OpenGl_MaterialState& aMatState = myShaderManager->MaterialState();
+  float anAlphaCutoff = (anAspect->AlphaMode() == Graphic3d_AlphaMode_Mask
                          || anAspect->AlphaMode() == Graphic3d_AlphaMode_MaskBlend)
-                                                ? anAspect->AlphaCutoff()
-                                                : ShortRealLast();
+                          ? anAspect->AlphaCutoff()
+                          : ShortRealLast();
   if (anAspect->ToDrawEdges())
   {
     if (anAspect->InteriorStyle() == Aspect_IS_EMPTY
diff --git c/src/OpenGl/OpenGl_GlFunctions.cxx w/src/OpenGl/OpenGl_GlFunctions.cxx
index 2529d5a06e..e71c8904d4 100644
--- c/src/OpenGl/OpenGl_GlFunctions.cxx
+++ w/src/OpenGl/OpenGl_GlFunctions.cxx
@@ -24,7 +24,8 @@
 #include <OpenGl_ArbTexBindless.hxx>
 #include <OpenGl_GlCore46.hxx>
 
-#if !defined(HAVE_EGL) && defined(HAVE_XLIB)
+#include <Standard_Macro_Runtime.hxx>
+#if (!defined(HAVE_EGL) && defined(HAVE_XLIB)) || WAYLAND_COMPAT
   #include <GL/glx.h>
 #endif
 
@@ -2285,12 +2286,12 @@ void OpenGl_GlFunctions::load(OpenGl_Context& theCtx, Standard_Boolean theIsCore
   theCtx.arbSampleShading  = checkExtensionShort("GL_ARB_sample_shading");
   theCtx.arbDepthClamp = isGlGreaterEqualShort(3, 2) || checkExtensionShort("GL_ARB_depth_clamp")
                          || checkExtensionShort("NV_depth_clamp");
-  theCtx.extBgra   = isGlGreaterEqualShort(1, 2) || checkExtensionShort("GL_EXT_bgra");
-  theCtx.extTexR16 = true;
-  theCtx.extAnis   = checkExtensionShort("GL_EXT_texture_filter_anisotropic");
-  theCtx.extPDS    = checkExtensionShort("GL_EXT_packed_depth_stencil");
-  theCtx.atiMem    = checkExtensionShort("GL_ATI_meminfo");
-  theCtx.nvxMem    = checkExtensionShort("GL_NVX_gpu_memory_info");
+  theCtx.extBgra       = isGlGreaterEqualShort(1, 2) || checkExtensionShort("GL_EXT_bgra");
+  theCtx.extTexR16     = true;
+  theCtx.extAnis       = checkExtensionShort("GL_EXT_texture_filter_anisotropic");
+  theCtx.extPDS        = checkExtensionShort("GL_EXT_packed_depth_stencil");
+  theCtx.atiMem        = checkExtensionShort("GL_ATI_meminfo");
+  theCtx.nvxMem        = checkExtensionShort("GL_NVX_gpu_memory_info");
 
   theCtx.hasDrawBuffers = isGlGreaterEqualShort(2, 0) ? OpenGl_FeatureInCore
                           : theCtx.arbDrawBuffers     ? OpenGl_FeatureInExtensions
@@ -2320,9 +2321,10 @@ void OpenGl_GlFunctions::load(OpenGl_Context& theCtx, Standard_Boolean theIsCore
        has46 = false;
 
   // retrieve platform-dependent extensions
-  #if defined(HAVE_EGL)
+  #ifdef COMPILE_EGL
     //
-  #elif defined(_WIN32)
+  #endif
+  #ifdef COMPILE_WIN32
   if (FindProcShort(wglGetExtensionsStringARB))
   {
     const char* aWglExts = this->wglGetExtensionsStringARB(wglGetCurrentDC());
@@ -2356,25 +2358,29 @@ void OpenGl_GlFunctions::load(OpenGl_Context& theCtx, Standard_Boolean theIsCore
       FindProcShort(wglGetContextGPUIDAMD);
     }
   }
-  #elif defined(HAVE_XLIB)
-  const char* aGlxExts = ::glXQueryExtensionsString((Display*)theCtx.myDisplay,
-                                                    DefaultScreen((Display*)theCtx.myDisplay));
-  if (checkExtensionShort(aGlxExts, "GLX_EXT_swap_control"))
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    FindProcShort(glXSwapIntervalEXT);
-  }
-  if (checkExtensionShort(aGlxExts, "GLX_SGI_swap_control"))
-  {
-    FindProcShort(glXSwapIntervalSGI);
-  }
-  if (checkExtensionShort(aGlxExts, "GLX_MESA_query_renderer"))
-  {
-    FindProcShort(glXQueryRendererIntegerMESA);
-    FindProcShort(glXQueryCurrentRendererIntegerMESA);
-    FindProcShort(glXQueryRendererStringMESA);
-    FindProcShort(glXQueryCurrentRendererStringMESA);
-  }
+    const char* aGlxExts = ::glXQueryExtensionsString((Display*)theCtx.myDisplay,
+                                                      DefaultScreen((Display*)theCtx.myDisplay));
+    if (checkExtensionShort(aGlxExts, "GLX_EXT_swap_control"))
+    {
+      FindProcShort(glXSwapIntervalEXT);
+    }
+    if (checkExtensionShort(aGlxExts, "GLX_SGI_swap_control"))
+    {
+      FindProcShort(glXSwapIntervalSGI);
+    }
+    if (checkExtensionShort(aGlxExts, "GLX_MESA_query_renderer"))
+    {
+      FindProcShort(glXQueryRendererIntegerMESA);
+      FindProcShort(glXQueryCurrentRendererIntegerMESA);
+      FindProcShort(glXQueryRendererStringMESA);
+      FindProcShort(glXQueryCurrentRendererStringMESA);
+    }
     // extSwapTear = checkExtensionShort (aGlxExts, "GLX_EXT_swap_control_tear");
+  }
   #endif
 
   // load OpenGL 1.2 new functions
diff --git c/src/OpenGl/OpenGl_GraphicDriver.cxx w/src/OpenGl/OpenGl_GraphicDriver.cxx
index a06c449019..7f317da1a3 100644
--- c/src/OpenGl/OpenGl_GraphicDriver.cxx
+++ w/src/OpenGl/OpenGl_GraphicDriver.cxx
@@ -72,11 +72,21 @@ IMPLEMENT_STANDARD_RTTIEXT(OpenGl_GraphicDriver, Graphic3d_GraphicDriver)
   #define OpenGl_USE_GLES2
 #endif
 
+#include <Standard_Macro_Runtime.hxx>
+#if WAYLAND_COMPAT
+  #include <Aspect_NeutralWindow.hxx>
+  #include <Xw_Window.hxx>
+  #include <X11/Xlib.h>
+  #include <X11/Xutil.h>
+  #include <EGL/egl.h>
+  #include <GL/glx.h>
+#endif
+
 namespace
 {
 static const Handle(OpenGl_Context) TheNullGlCtx;
 
-#if defined(HAVE_EGL)
+#if defined(HAVE_EGL) || WAYLAND_COMPAT
 //! Wrapper over eglChooseConfig() called with preferred defaults.
 static EGLConfig chooseEglSurfConfig(EGLDisplay theDisplay, const Handle(OpenGl_Caps)& theCaps)
 {
@@ -149,7 +159,8 @@ static EGLConfig chooseEglSurfConfig(EGLDisplay theDisplay, const Handle(OpenGl_
   }
   return aCfg;
 }
-#elif defined(HAVE_XLIB)
+#endif
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
 //! Search for RGBA double-buffered visual with stencil buffer.
 static int TheDoubleBuffVisual[] =
   {GLX_RGBA, GLX_DEPTH_SIZE, 16, GLX_STENCIL_SIZE, 1, GLX_DOUBLEBUFFER, None};
@@ -187,25 +198,31 @@ OpenGl_GraphicDriver::OpenGl_GraphicDriver(const Handle(Aspect_DisplayConnection
       myMapOfView(1, NCollection_BaseAllocator::CommonBaseAllocator()),
       myMapOfStructure(1, NCollection_BaseAllocator::CommonBaseAllocator())
 {
-#if defined(HAVE_EGL)
-  myEglDisplay = (Aspect_Display)EGL_NO_DISPLAY;
-  myEglContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
+#if defined(HAVE_EGL) || WAYLAND_COMPAT
+  if (!RUN_UNIX_X11)
+  {
+    myEglDisplay = (Aspect_Display)EGL_NO_DISPLAY;
+    myEglContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
+  }
 #endif
 #if defined(OpenGl_USE_GLES2)
   myCaps->contextCompatible = false;
 #endif
 
-#if defined(HAVE_XLIB)
-  if (myDisplayConnection.IsNull())
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (!RUN_WAYLAND)
   {
-    // throw Aspect_GraphicDeviceDefinitionError("OpenGl_GraphicDriver: cannot connect to X
-    // server!");
-    return;
-  }
+    if (myDisplayConnection.IsNull())
+    {
+      // throw Aspect_GraphicDeviceDefinitionError("OpenGl_GraphicDriver: cannot connect to X
+      // server!");
+      return;
+    }
 
-  Display* aDisplay = (Display*)myDisplayConnection->GetDisplayAspect();
-  Bool     toSync   = ::getenv("CSF_GraphicSync") != NULL || ::getenv("CALL_SYNCHRO_X") != NULL;
-  XSynchronize(aDisplay, toSync);
+    Display* aDisplay = (Display*)myDisplayConnection->GetDisplayAspect();
+    Bool     toSync   = ::getenv("CSF_GraphicSync") != NULL || ::getenv("CALL_SYNCHRO_X") != NULL;
+    XSynchronize(aDisplay, toSync);
+  }
 #endif
   if (theToInitialize && !InitContext())
   {
@@ -282,31 +299,34 @@ void OpenGl_GraphicDriver::ReleaseContext()
     aWindow->GetGlContext()->forcedRelease();
   }
 
-#if defined(HAVE_EGL)
-  if (myIsOwnContext)
+#if defined(HAVE_EGL) || WAYLAND_COMPAT
+  if (!RUN_UNIX_X11)
   {
-    if (myEglContext != (Aspect_RenderingContext)EGL_NO_CONTEXT)
+    if (myIsOwnContext)
     {
-      if (eglMakeCurrent((EGLDisplay)myEglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT)
-          != EGL_TRUE)
+      if (myEglContext != (Aspect_RenderingContext)EGL_NO_CONTEXT)
       {
-        ::Message::SendWarning("OpenGl_GraphicDriver, FAILED to release OpenGL context");
+        if (eglMakeCurrent((EGLDisplay)myEglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT)
+            != EGL_TRUE)
+        {
+          ::Message::SendWarning("OpenGl_GraphicDriver, FAILED to release OpenGL context");
+        }
+        eglDestroyContext((EGLDisplay)myEglDisplay, (EGLContext)myEglContext);
+      }
+
+      if (myEglDisplay != (Aspect_Display)EGL_NO_DISPLAY)
+      {
+        if (eglTerminate((EGLDisplay)myEglDisplay) != EGL_TRUE)
+        {
+          ::Message::SendWarning("OpenGl_GraphicDriver, EGL, eglTerminate FAILED");
+        }
       }
-      eglDestroyContext((EGLDisplay)myEglDisplay, (EGLContext)myEglContext);
     }
 
-    if (myEglDisplay != (Aspect_Display)EGL_NO_DISPLAY)
-    {
-      if (eglTerminate((EGLDisplay)myEglDisplay) != EGL_TRUE)
-      {
-        ::Message::SendWarning("OpenGl_GraphicDriver, EGL, eglTerminate FAILED");
-      }
-    }
+    myEglDisplay = (Aspect_Display)EGL_NO_DISPLAY;
+    myEglContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
+    myEglConfig  = NULL;
   }
-
-  myEglDisplay = (Aspect_Display)EGL_NO_DISPLAY;
-  myEglContext = (Aspect_RenderingContext)EGL_NO_CONTEXT;
-  myEglConfig  = NULL;
 #endif
   myIsOwnContext = Standard_False;
 }
@@ -316,80 +336,85 @@ void OpenGl_GraphicDriver::ReleaseContext()
 Standard_Boolean OpenGl_GraphicDriver::InitContext()
 {
   ReleaseContext();
-#if defined(HAVE_EGL)
-
-  #if defined(HAVE_XLIB)
-  if (myDisplayConnection.IsNull())
+#if defined(HAVE_EGL) || WAYLAND_COMPAT
+  do
   {
-    return Standard_False;
-  }
-  Display* aDisplay = (Display*)myDisplayConnection->GetDisplayAspect();
-  myEglDisplay      = (Aspect_Display)eglGetDisplay(aDisplay);
+    if (RUN_UNIX_X11)
+      break;
+
+  #if defined(HAVE_XLIB) && !WAYLAND_COMPAT
+    if (myDisplayConnection.IsNull())
+    {
+      return Standard_False;
+    }
+    Display* aDisplay = (Display*)myDisplayConnection->GetDisplayAspect();
+    myEglDisplay      = (Aspect_Display)eglGetDisplay(aDisplay);
   #else
-  myEglDisplay = (Aspect_Display)eglGetDisplay(EGL_DEFAULT_DISPLAY);
+    myEglDisplay = (Aspect_Display)eglGetDisplay(EGL_DEFAULT_DISPLAY);
   #endif
-  if ((EGLDisplay)myEglDisplay == EGL_NO_DISPLAY)
-  {
-    ::Message::SendFail("Error: no EGL display");
-    return Standard_False;
-  }
+    if ((EGLDisplay)myEglDisplay == EGL_NO_DISPLAY)
+    {
+      ::Message::SendFail("Error: no EGL display");
+      return Standard_False;
+    }
 
-  EGLint aVerMajor = 0;
-  EGLint aVerMinor = 0;
-  if (eglInitialize((EGLDisplay)myEglDisplay, &aVerMajor, &aVerMinor) != EGL_TRUE)
-  {
-    ::Message::SendFail("Error: EGL display is unavailable");
-    return Standard_False;
-  }
+    EGLint aVerMajor = 0;
+    EGLint aVerMinor = 0;
+    if (eglInitialize((EGLDisplay)myEglDisplay, &aVerMajor, &aVerMinor) != EGL_TRUE)
+    {
+      ::Message::SendFail("Error: EGL display is unavailable");
+      return Standard_False;
+    }
 
-  myEglConfig = chooseEglSurfConfig((EGLDisplay)myEglDisplay, myCaps);
-  if (myEglConfig == NULL)
-  {
-    ::Message::SendFail("Error: EGL does not provide compatible configurations");
-    return Standard_False;
-  }
+    myEglConfig = chooseEglSurfConfig((EGLDisplay)myEglDisplay, myCaps);
+    if (myEglConfig == NULL)
+    {
+      ::Message::SendFail("Error: EGL does not provide compatible configurations");
+      return Standard_False;
+    }
 
   #if defined(OpenGl_USE_GLES2)
-  EGLint anEglCtxAttribs3[] = {EGL_CONTEXT_CLIENT_VERSION, 3, EGL_NONE, EGL_NONE};
-  EGLint anEglCtxAttribs2[] = {EGL_CONTEXT_CLIENT_VERSION, 2, EGL_NONE, EGL_NONE};
-  if (eglBindAPI(EGL_OPENGL_ES_API) != EGL_TRUE)
-  {
-    ::Message::SendFail("Error: EGL does not provide OpenGL ES client");
-    return Standard_False;
-  }
-  if (myCaps->contextMajorVersionUpper != 2)
-  {
-    myEglContext = (Aspect_RenderingContext)
-      eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs3);
-  }
-  if ((EGLContext)myEglContext == EGL_NO_CONTEXT)
-  {
-    myEglContext = (Aspect_RenderingContext)
-      eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs2);
-  }
+    EGLint anEglCtxAttribs3[] = {EGL_CONTEXT_CLIENT_VERSION, 3, EGL_NONE, EGL_NONE};
+    EGLint anEglCtxAttribs2[] = {EGL_CONTEXT_CLIENT_VERSION, 2, EGL_NONE, EGL_NONE};
+    if (eglBindAPI(EGL_OPENGL_ES_API) != EGL_TRUE)
+    {
+      ::Message::SendFail("Error: EGL does not provide OpenGL ES client");
+      return Standard_False;
+    }
+    if (myCaps->contextMajorVersionUpper != 2)
+    {
+      myEglContext = (Aspect_RenderingContext)
+        eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs3);
+    }
+    if ((EGLContext)myEglContext == EGL_NO_CONTEXT)
+    {
+      myEglContext = (Aspect_RenderingContext)
+        eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs2);
+    }
   #else
-  EGLint* anEglCtxAttribs = NULL;
-  if (eglBindAPI(EGL_OPENGL_API) != EGL_TRUE)
-  {
-    ::Message::SendFail("Error: EGL does not provide OpenGL client");
-    return Standard_False;
-  }
-  myEglContext = (Aspect_RenderingContext)
-    eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs);
+    EGLint* anEglCtxAttribs = NULL;
+    if (eglBindAPI(EGL_OPENGL_API) != EGL_TRUE)
+    {
+      ::Message::SendFail("Error: EGL does not provide OpenGL client");
+      return Standard_False;
+    }
+    myEglContext = (Aspect_RenderingContext)
+      eglCreateContext((EGLDisplay)myEglDisplay, myEglConfig, EGL_NO_CONTEXT, anEglCtxAttribs);
   #endif
 
-  if ((EGLContext)myEglContext == EGL_NO_CONTEXT)
-  {
-    ::Message::SendFail("Error: EGL is unable to create OpenGL context");
-    return Standard_False;
-  }
-  // eglMakeCurrent() fails or even crash with EGL_NO_SURFACE on some implementations
-  // if (eglMakeCurrent ((EGLDisplay )myEglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, (EGLContext
-  // )myEglContext) != EGL_TRUE)
-  //{
-  //  ::Message::SendFail ("Error: EGL is unable bind OpenGL context");
-  //  return Standard_False;
-  //}
+    if ((EGLContext)myEglContext == EGL_NO_CONTEXT)
+    {
+      ::Message::SendFail("Error: EGL is unable to create OpenGL context");
+      return Standard_False;
+    }
+    // eglMakeCurrent() fails or even crash with EGL_NO_SURFACE on some implementations
+    // if (eglMakeCurrent ((EGLDisplay )myEglDisplay, EGL_NO_SURFACE, EGL_NO_SURFACE, (EGLContext
+    // )myEglContext) != EGL_TRUE)
+    //{
+    //  ::Message::SendFail ("Error: EGL is unable bind OpenGL context");
+    //  return Standard_False;
+    //}
+  } while (false);
 #endif
   chooseVisualInfo();
   myIsOwnContext = Standard_True;
@@ -403,32 +428,38 @@ Standard_Boolean OpenGl_GraphicDriver::InitEglContext(Aspect_Display          th
                                                       void*                   theEglConfig)
 {
   ReleaseContext();
-#if defined(HAVE_EGL)
-  #if defined(HAVE_XLIB)
-  if (myDisplayConnection.IsNull())
+#if defined(HAVE_EGL) || WAYLAND_COMPAT
+  do
   {
-    return Standard_False;
-  }
-  #endif
+    if (RUN_UNIX_X11)
+      break;
 
-  if ((EGLDisplay)theEglDisplay == EGL_NO_DISPLAY || (EGLContext)theEglContext == EGL_NO_CONTEXT)
-  {
-    return Standard_False;
-  }
-  myEglDisplay = theEglDisplay;
-  myEglContext = theEglContext;
-  myEglConfig  = theEglConfig;
-  if (theEglConfig == NULL)
-  {
-    myEglConfig = chooseEglSurfConfig((EGLDisplay)myEglDisplay, myCaps);
-    if (myEglConfig == NULL)
+  #if defined(HAVE_XLIB) && !WAYLAND_COMPAT
+    if (myDisplayConnection.IsNull())
     {
-      ::Message::SendFail("Error: EGL does not provide compatible configurations");
       return Standard_False;
     }
-  }
-  chooseVisualInfo();
-  return Standard_True;
+  #endif
+
+    if ((EGLDisplay)theEglDisplay == EGL_NO_DISPLAY || (EGLContext)theEglContext == EGL_NO_CONTEXT)
+    {
+      return Standard_False;
+    }
+    myEglDisplay = theEglDisplay;
+    myEglContext = theEglContext;
+    myEglConfig  = theEglConfig;
+    if (theEglConfig == NULL)
+    {
+      myEglConfig = chooseEglSurfConfig((EGLDisplay)myEglDisplay, myCaps);
+      if (myEglConfig == NULL)
+      {
+        ::Message::SendFail("Error: EGL does not provide compatible configurations");
+        return Standard_False;
+      }
+    }
+    chooseVisualInfo();
+    return Standard_True;
+  } while (false);
 #else
   (void)theEglDisplay;
   (void)theEglContext;
@@ -446,12 +477,15 @@ void OpenGl_GraphicDriver::chooseVisualInfo()
     return;
   }
 
-#if defined(HAVE_XLIB)
-  Display* aDisp = (Display*)myDisplayConnection->GetDisplayAspect();
+#if defined(HAVE_XLIB) || WAYLAND_COMPAT
+  if (RUN_WAYLAND)
+    return;
 
+  Display*        aDisp      = (Display*)myDisplayConnection->GetDisplayAspect();
   XVisualInfo*    aVisInfo   = NULL;
   Aspect_FBConfig anFBConfig = NULL;
-  #if defined(HAVE_EGL)
+
+  #if defined(HAVE_EGL) && !WAYLAND_COMPAT
   XVisualInfo aVisInfoTmp;
   memset(&aVisInfoTmp, 0, sizeof(aVisInfoTmp));
   aVisInfoTmp.screen = DefaultScreen(aDisp);
diff --git c/src/OpenGl/OpenGl_Window.cxx w/src/OpenGl/OpenGl_Window.cxx
index 7382d9b75c..97f27875a4 100644
--- c/src/OpenGl/OpenGl_Window.cxx
+++ w/src/OpenGl/OpenGl_Window.cxx
@@ -34,7 +34,11 @@
 
 IMPLEMENT_STANDARD_RTTIEXT(OpenGl_Window, Standard_Transient)
 
-#if defined(HAVE_EGL)
+#include <Standard_Macro_Runtime.hxx>
+#if WAYLAND_COMPAT
+  #include <EGL/egl.h>
+  #include <GL/glx.h>
+#elif defined(HAVE_EGL)
   #include <EGL/egl.h>
 #elif defined(HAVE_XLIB)
   #include <GL/glx.h>
@@ -45,10 +49,10 @@ IMPLEMENT_STANDARD_RTTIEXT(OpenGl_Window, Standard_Transient)
 namespace
 {
 
-  #if defined(HAVE_EGL)
+  #ifdef COMPILE_EGL
   //
-  #elif defined(_WIN32)
-
+  #endif
+  #ifdef COMPILE_WIN32
   // WGL_ARB_pixel_format
     #ifndef WGL_NUMBER_PIXEL_FORMATS_ARB
       #define WGL_NUMBER_PIXEL_FORMATS_ARB 0x2000
@@ -126,8 +130,8 @@ static LRESULT CALLBACK wndProcDummy(HWND theWin, UINT theMsg, WPARAM theParamW,
 {
   return DefWindowProcW(theWin, theMsg, theParamW, theParamL);
 }
-  #elif defined(HAVE_XLIB)
-
+  #endif
+  #ifdef COMPILE_XLIB
   // GLX_ARB_create_context
     #ifndef GLX_CONTEXT_MAJOR_VERSION_ARB
       #define GLX_CONTEXT_DEBUG_BIT_ARB 0x00000001
@@ -141,7 +145,6 @@ static LRESULT CALLBACK wndProcDummy(HWND theWin, UINT theMsg, WPARAM theParamW,
       #define GLX_CONTEXT_COMPATIBILITY_PROFILE_BIT_ARB 0x00000002
       #define GLX_CONTEXT_PROFILE_MASK_ARB 0x9126
     #endif
-
 //! Dummy XError handler which just skips errors
 static int xErrorDummyHandler(Display* /*theDisplay*/, XErrorEvent* /*theErrorEvent*/)
 {
@@ -190,118 +193,122 @@ void OpenGl_Window::Init(const Handle(OpenGl_GraphicDriver)& theDriver,
 
   Standard_Boolean isCoreProfile = Standard_False;
 
-  #if defined(HAVE_EGL)
-  EGLDisplay anEglDisplay = (EGLDisplay)theDriver->getRawGlDisplay();
-  EGLContext anEglContext = (EGLContext)theDriver->getRawGlContext();
-  EGLConfig  anEglConfig  = (EGLConfig)theDriver->getRawGlConfig();
-  if (anEglDisplay == EGL_NO_DISPLAY || anEglContext == EGL_NO_CONTEXT
-      || (anEglConfig == NULL && (EGLContext)theGContext == EGL_NO_CONTEXT))
+  #ifdef COMPILE_EGL
+  if (!RUN_UNIX_X11)
   {
-    throw Aspect_GraphicDeviceDefinitionError(
-      "OpenGl_Window, EGL does not provide compatible configurations!");
-  }
-
-  EGLSurface anEglSurf = EGL_NO_SURFACE;
-  if ((EGLContext)theGContext == EGL_NO_CONTEXT)
-  {
-    // EGL_KHR_gl_colorspace extension specifies if OpenGL should write into window buffer as into
-    // sRGB or RGB framebuffer
-    // const int aSurfAttribs[] =
-    //{
-    //  EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ? EGL_GL_COLORSPACE_SRGB_KHR :
-    //  EGL_GL_COLORSPACE_LINEAR_KHR, EGL_NONE,
-    //};
-
-    // create new surface
-    anEglSurf = eglCreateWindowSurface(anEglDisplay,
-                                       anEglConfig,
-                                       (EGLNativeWindowType)myPlatformWindow->NativeHandle(),
-                                       NULL);
-    if (anEglSurf == EGL_NO_SURFACE && myPlatformWindow->NativeHandle() != 0)
+    EGLDisplay anEglDisplay = (EGLDisplay)theDriver->getRawGlDisplay();
+    EGLContext anEglContext = (EGLContext)theDriver->getRawGlContext();
+    EGLConfig  anEglConfig  = (EGLConfig)theDriver->getRawGlConfig();
+    if (anEglDisplay == EGL_NO_DISPLAY || anEglContext == EGL_NO_CONTEXT
+        || (anEglConfig == NULL && (EGLContext)theGContext == EGL_NO_CONTEXT))
     {
       throw Aspect_GraphicDeviceDefinitionError(
-        "OpenGl_Window, EGL is unable to create surface for window!");
+        "OpenGl_Window, EGL does not provide compatible configurations!");
     }
-    else if (anEglSurf == EGL_NO_SURFACE)
-    {
-      // window-less EGL context (off-screen)
-      // throw Aspect_GraphicDeviceDefinitionError("OpenGl_Window, EGL is unable to retrieve current
-      // surface!");
-      if (anEglConfig != NULL)
-      {
-    #if !defined(__EMSCRIPTEN__) // eglCreatePbufferSurface() is not implemented by Emscripten EGL
-        const int aSurfAttribs[] = {EGL_WIDTH,
-                                    mySize.x(),
-                                    EGL_HEIGHT,
-                                    mySize.y(),
-                                    // EGL_KHR_gl_colorspace extension specifies if OpenGL should
-                                    // write into window buffer as into sRGB or RGB framebuffer
-                                    // EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ?
-                                    // EGL_GL_COLORSPACE_SRGB_KHR : EGL_GL_COLORSPACE_LINEAR_KHR,
-                                    EGL_NONE};
-        anEglSurf                = eglCreatePbufferSurface(anEglDisplay, anEglConfig, aSurfAttribs);
-        if (anEglSurf == EGL_NO_SURFACE)
-        {
-          throw Aspect_GraphicDeviceDefinitionError(
-            "OpenGl_Window, EGL is unable to create off-screen surface!");
-        }
-    #endif
-      }
-      myGlContext->PushMessage(
-        GL_DEBUG_SOURCE_APPLICATION,
-        GL_DEBUG_TYPE_PORTABILITY,
-        0,
-        GL_DEBUG_SEVERITY_LOW,
-        "OpenGl_Window::CreateWindow: WARNING, a Window is created without a EGL Surface!");
-    }
-  }
-  else if (theGContext != anEglContext)
-  {
-    throw Aspect_GraphicDeviceDefinitionError(
-      "OpenGl_Window, EGL is used in unsupported combination!");
-  }
-  else
-  {
-    anEglSurf = eglGetCurrentSurface(EGL_DRAW);
-    if (anEglSurf == EGL_NO_SURFACE)
-    {
-      // window-less EGL context (off-screen)
-      // throw Aspect_GraphicDeviceDefinitionError("OpenGl_Window, EGL is unable to retrieve current
-      // surface!");
-      if (anEglConfig != NULL)
-      {
-    #if !defined(__EMSCRIPTEN__) // eglCreatePbufferSurface() is not implemented by Emscripten EGL
-        const int aSurfAttribs[] = {EGL_WIDTH,
-                                    mySize.x(),
-                                    EGL_HEIGHT,
-                                    mySize.y(),
-                                    // EGL_KHR_gl_colorspace extension specifies if OpenGL should
-                                    // write into window buffer as into sRGB or RGB framebuffer
-                                    // EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ?
-                                    // EGL_GL_COLORSPACE_SRGB_KHR : EGL_GL_COLORSPACE_LINEAR_KHR,
-                                    EGL_NONE};
-        anEglSurf                = eglCreatePbufferSurface(anEglDisplay, anEglConfig, aSurfAttribs);
-        if (anEglSurf == EGL_NO_SURFACE)
-        {
-          throw Aspect_GraphicDeviceDefinitionError(
-            "OpenGl_Window, EGL is unable to create off-screen surface!");
-        }
-    #endif
-      }
-      myGlContext->PushMessage(
-        GL_DEBUG_SOURCE_APPLICATION,
-        GL_DEBUG_TYPE_PORTABILITY,
-        0,
-        GL_DEBUG_SEVERITY_LOW,
-        "OpenGl_Window::CreateWindow: WARNING, a Window is created without a EGL Surface!");
-    }
-  }
 
-  myGlContext->Init((Aspect_Drawable)anEglSurf,
-                    (Aspect_Display)anEglDisplay,
-                    (Aspect_RenderingContext)anEglContext,
-                    isCoreProfile);
-  #elif defined(_WIN32)
+    EGLSurface anEglSurf = EGL_NO_SURFACE;
+    if ((EGLContext)theGContext == EGL_NO_CONTEXT)
+    {
+      // EGL_KHR_gl_colorspace extension specifies if OpenGL should write into window buffer as into
+      // sRGB or RGB framebuffer
+      // const int aSurfAttribs[] =
+      //{
+      //  EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ? EGL_GL_COLORSPACE_SRGB_KHR :
+      //  EGL_GL_COLORSPACE_LINEAR_KHR, EGL_NONE,
+      //};
+
+      // create new surface
+      anEglSurf = eglCreateWindowSurface(anEglDisplay,
+                                         anEglConfig,
+                                         (EGLNativeWindowType)myPlatformWindow->NativeHandle(),
+                                         NULL);
+      if (anEglSurf == EGL_NO_SURFACE && myPlatformWindow->NativeHandle() != 0)
+      {
+        throw Aspect_GraphicDeviceDefinitionError(
+          "OpenGl_Window, EGL is unable to create surface for window!");
+      }
+      else if (anEglSurf == EGL_NO_SURFACE)
+      {
+        // window-less EGL context (off-screen)
+        // throw Aspect_GraphicDeviceDefinitionError("OpenGl_Window, EGL is unable to retrieve
+        // current surface!");
+        if (anEglConfig != NULL)
+        {
+    #if !defined(__EMSCRIPTEN__) // eglCreatePbufferSurface() is not implemented by Emscripten EGL
+          const int aSurfAttribs[] = {EGL_WIDTH,
+                                      mySize.x(),
+                                      EGL_HEIGHT,
+                                      mySize.y(),
+                                      // EGL_KHR_gl_colorspace extension specifies if OpenGL should
+                                      // write into window buffer as into sRGB or RGB framebuffer
+                                      // EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ?
+                                      // EGL_GL_COLORSPACE_SRGB_KHR : EGL_GL_COLORSPACE_LINEAR_KHR,
+                                      EGL_NONE};
+          anEglSurf = eglCreatePbufferSurface(anEglDisplay, anEglConfig, aSurfAttribs);
+          if (anEglSurf == EGL_NO_SURFACE)
+          {
+            throw Aspect_GraphicDeviceDefinitionError(
+              "OpenGl_Window, EGL is unable to create off-screen surface!");
+          }
+    #endif
+        }
+        myGlContext->PushMessage(
+          GL_DEBUG_SOURCE_APPLICATION,
+          GL_DEBUG_TYPE_PORTABILITY,
+          0,
+          GL_DEBUG_SEVERITY_LOW,
+          "OpenGl_Window::CreateWindow: WARNING, a Window is created without a EGL Surface!");
+      }
+    }
+    else if (theGContext != anEglContext)
+    {
+      throw Aspect_GraphicDeviceDefinitionError(
+        "OpenGl_Window, EGL is used in unsupported combination!");
+    }
+    else
+    {
+      anEglSurf = eglGetCurrentSurface(EGL_DRAW);
+      if (anEglSurf == EGL_NO_SURFACE)
+      {
+        // window-less EGL context (off-screen)
+        // throw Aspect_GraphicDeviceDefinitionError("OpenGl_Window, EGL is unable to retrieve
+        // current surface!");
+        if (anEglConfig != NULL)
+        {
+    #if !defined(__EMSCRIPTEN__) // eglCreatePbufferSurface() is not implemented by Emscripten EGL
+          const int aSurfAttribs[] = {EGL_WIDTH,
+                                      mySize.x(),
+                                      EGL_HEIGHT,
+                                      mySize.y(),
+                                      // EGL_KHR_gl_colorspace extension specifies if OpenGL should
+                                      // write into window buffer as into sRGB or RGB framebuffer
+                                      // EGL_GL_COLORSPACE_KHR, !theCaps->sRGBDisable ?
+                                      // EGL_GL_COLORSPACE_SRGB_KHR : EGL_GL_COLORSPACE_LINEAR_KHR,
+                                      EGL_NONE};
+          anEglSurf = eglCreatePbufferSurface(anEglDisplay, anEglConfig, aSurfAttribs);
+          if (anEglSurf == EGL_NO_SURFACE)
+          {
+            throw Aspect_GraphicDeviceDefinitionError(
+              "OpenGl_Window, EGL is unable to create off-screen surface!");
+          }
+    #endif
+        }
+        myGlContext->PushMessage(
+          GL_DEBUG_SOURCE_APPLICATION,
+          GL_DEBUG_TYPE_PORTABILITY,
+          0,
+          GL_DEBUG_SEVERITY_LOW,
+          "OpenGl_Window::CreateWindow: WARNING, a Window is created without a EGL Surface!");
+      }
+    }
+
+    myGlContext->Init((Aspect_Drawable)anEglSurf,
+                      (Aspect_Display)anEglDisplay,
+                      (Aspect_RenderingContext)anEglContext,
+                      isCoreProfile);
+  }
+  #endif
+  #ifdef COMPILE_WIN32
   (void)theDriver;
   HWND  aWindow   = (HWND)myPlatformWindow->NativeHandle();
   HDC   aWindowDC = GetDC(aWindow);
@@ -601,156 +608,161 @@ void OpenGl_Window::Init(const Handle(OpenGl_GraphicDriver)& theDriver,
                     (Aspect_Handle)aWindowDC,
                     (Aspect_RenderingContext)aGContext,
                     isCoreProfile);
-  #elif defined(HAVE_XLIB)
-  Window     aWindow   = (Window)myPlatformWindow->NativeHandle();
-  Display*   aDisp     = (Display*)theDriver->GetDisplayConnection()->GetDisplayAspect();
-  GLXContext aGContext = (GLXContext)theGContext;
-  GLXContext aSlaveCtx = !theShareCtx.IsNull() ? (GLXContext)theShareCtx->myGContext : NULL;
-
-  XWindowAttributes aWinAttribs;
-  XGetWindowAttributes(aDisp, aWindow, &aWinAttribs);
-  XVisualInfo aVisInfo;
-  aVisInfo.visualid = aWinAttribs.visual->visualid;
-  aVisInfo.screen   = DefaultScreen(aDisp);
-  int                                          aNbItems;
-  std::unique_ptr<XVisualInfo, int (*)(void*)> aVis(
-    XGetVisualInfo(aDisp, VisualIDMask | VisualScreenMask, &aVisInfo, &aNbItems),
-    &XFree);
-  int isGl = 0;
-  if (aVis.get() == NULL)
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    throw Aspect_GraphicDeviceDefinitionError(
-      "OpenGl_Window::CreateWindow: XGetVisualInfo is unable to choose needed configuration in "
-      "existing OpenGL context. ");
-  }
-  else if (glXGetConfig(aDisp, aVis.get(), GLX_USE_GL, &isGl) != 0 || !isGl)
-  {
-    throw Aspect_GraphicDeviceDefinitionError(
-      "OpenGl_Window::CreateWindow: window Visual does not support GL rendering!");
-  }
+    Window     aWindow   = (Window)myPlatformWindow->NativeHandle();
+    Display*   aDisp     = (Display*)theDriver->GetDisplayConnection()->GetDisplayAspect();
+    GLXContext aGContext = (GLXContext)theGContext;
+    GLXContext aSlaveCtx = !theShareCtx.IsNull() ? (GLXContext)theShareCtx->myGContext : NULL;
 
-  // create new context
-  GLXFBConfig anFBConfig = myPlatformWindow->NativeFBConfig();
-  const char* aGlxExts   = glXQueryExtensionsString(aDisp, aVisInfo.screen);
-  if (myOwnGContext && anFBConfig != NULL
-      && OpenGl_Context::CheckExtension(aGlxExts, "GLX_ARB_create_context_profile"))
-  {
-    // Replace default XError handler to ignore errors.
-    // Warning - this is global for all threads!
-    typedef int (*xerrorhandler_t)(Display*, XErrorEvent*);
-    xerrorhandler_t anOldHandler = XSetErrorHandler(xErrorDummyHandler);
-
-    typedef GLXContext (*glXCreateContextAttribsARB_t)(Display*    dpy,
-                                                       GLXFBConfig config,
-                                                       GLXContext  share_context,
-                                                       Bool        direct,
-                                                       const int*  attrib_list);
-    glXCreateContextAttribsARB_t aCreateCtxProc =
-      (glXCreateContextAttribsARB_t)glXGetProcAddress((const GLubyte*)"glXCreateContextAttribsARB");
-    if (!theCaps->contextCompatible)
-    {
-      int aCoreCtxAttribs[] = {GLX_CONTEXT_MAJOR_VERSION_ARB,
-                               3,
-                               GLX_CONTEXT_MINOR_VERSION_ARB,
-                               2,
-                               GLX_CONTEXT_PROFILE_MASK_ARB,
-                               GLX_CONTEXT_CORE_PROFILE_BIT_ARB,
-                               GLX_CONTEXT_FLAGS_ARB,
-                               theCaps->contextDebug ? GLX_CONTEXT_DEBUG_BIT_ARB : 0,
-                               0,
-                               0};
-
-      // try to create the core profile of highest OpenGL version supported by OCCT
-      for (int aLowVer4 = 6; aLowVer4 >= 0 && aGContext == NULL; --aLowVer4)
-      {
-        aCoreCtxAttribs[1] = 4;
-        aCoreCtxAttribs[3] = aLowVer4;
-        aGContext          = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCoreCtxAttribs);
-      }
-      for (int aLowVer3 = 3; aLowVer3 >= 2 && aGContext == NULL; --aLowVer3)
-      {
-        aCoreCtxAttribs[1] = 3;
-        aCoreCtxAttribs[3] = aLowVer3;
-        aGContext          = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCoreCtxAttribs);
-      }
-      isCoreProfile = aGContext != NULL;
-    }
-
-    if (aGContext == NULL)
-    {
-      int aCtxAttribs[] = {GLX_CONTEXT_FLAGS_ARB,
-                           theCaps->contextDebug ? GLX_CONTEXT_DEBUG_BIT_ARB : 0,
-                           0,
-                           0};
-      isCoreProfile     = Standard_False;
-      aGContext         = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCtxAttribs);
-
-      if (aGContext != NULL && !theCaps->contextCompatible)
-      {
-        TCollection_ExtendedString aMsg(
-          "OpenGl_Window::CreateWindow: core profile creation failed.");
-        myGlContext->PushMessage(GL_DEBUG_SOURCE_APPLICATION,
-                                 GL_DEBUG_TYPE_PORTABILITY,
-                                 0,
-                                 GL_DEBUG_SEVERITY_LOW,
-                                 aMsg);
-      }
-    }
-    XSetErrorHandler(anOldHandler);
-  }
-
-  if (myOwnGContext && aGContext == NULL)
-  {
-    aGContext = glXCreateContext(aDisp, aVis.get(), aSlaveCtx, GL_TRUE);
-    if (aGContext == NULL)
+    XWindowAttributes aWinAttribs;
+    XGetWindowAttributes(aDisp, aWindow, &aWinAttribs);
+    XVisualInfo aVisInfo;
+    aVisInfo.visualid = aWinAttribs.visual->visualid;
+    aVisInfo.screen   = DefaultScreen(aDisp);
+    int                                          aNbItems;
+    std::unique_ptr<XVisualInfo, int (*)(void*)> aVis(
+      XGetVisualInfo(aDisp, VisualIDMask | VisualScreenMask, &aVisInfo, &aNbItems),
+      &XFree);
+    int isGl = 0;
+    if (aVis.get() == NULL)
     {
       throw Aspect_GraphicDeviceDefinitionError(
-        "OpenGl_Window::CreateWindow: glXCreateContext failed.");
+        "OpenGl_Window::CreateWindow: XGetVisualInfo is unable to choose needed configuration in "
+        "existing OpenGL context. ");
+    }
+    else if (glXGetConfig(aDisp, aVis.get(), GLX_USE_GL, &isGl) != 0 || !isGl)
+    {
+      throw Aspect_GraphicDeviceDefinitionError(
+        "OpenGl_Window::CreateWindow: window Visual does not support GL rendering!");
     }
-  }
 
-  // check Visual for OpenGl context's parameters compatibility
-  TCollection_ExtendedString aList;
-  int                        isDoubleBuffer = 0, isRGBA = 0, isStereo = 0;
-  int                        aDepthSize = 0, aStencilSize = 0;
-  glXGetConfig(aDisp, aVis.get(), GLX_RGBA, &isRGBA);
-  glXGetConfig(aDisp, aVis.get(), GLX_DOUBLEBUFFER, &isDoubleBuffer);
-  glXGetConfig(aDisp, aVis.get(), GLX_STEREO, &isStereo);
-  glXGetConfig(aDisp, aVis.get(), GLX_DEPTH_SIZE, &aDepthSize);
-  glXGetConfig(aDisp, aVis.get(), GLX_STENCIL_SIZE, &aStencilSize);
-  if (aDepthSize < 1)
-    addMsgToList(aList, "no depth buffer");
-  if (aStencilSize < 1)
-    addMsgToList(aList, "no stencil buffer");
-  if (isRGBA == 0)
-    addMsgToList(aList, "no RGBA color buffer");
-  if (isDoubleBuffer == 0)
-    addMsgToList(aList, "no Double Buffer");
-  if (theCaps->contextStereo && isStereo == 0)
-  {
-    addMsgToList(aList, "no Quad Buffer");
-  }
-  else if (!theCaps->contextStereo && isStereo == 1)
-  {
-    addMsgToList(aList, "extra Quad Buffer");
-  }
-  if (!aList.IsEmpty())
-  {
-    TCollection_ExtendedString aMsg =
-      TCollection_ExtendedString("OpenGl_Window::CreateWindow: window Visual is incomplete: ")
-      + aList;
-    myGlContext->PushMessage(GL_DEBUG_SOURCE_APPLICATION,
-                             GL_DEBUG_TYPE_OTHER,
+    // create new context
+    GLXFBConfig anFBConfig = myPlatformWindow->NativeFBConfig();
+    const char* aGlxExts   = glXQueryExtensionsString(aDisp, aVisInfo.screen);
+    if (myOwnGContext && anFBConfig != NULL
+        && OpenGl_Context::CheckExtension(aGlxExts, "GLX_ARB_create_context_profile"))
+    {
+      // Replace default XError handler to ignore errors.
+      // Warning - this is global for all threads!
+      typedef int (*xerrorhandler_t)(Display*, XErrorEvent*);
+      xerrorhandler_t anOldHandler = XSetErrorHandler(xErrorDummyHandler);
+
+      typedef GLXContext (*glXCreateContextAttribsARB_t)(Display*    dpy,
+                                                         GLXFBConfig config,
+                                                         GLXContext  share_context,
+                                                         Bool        direct,
+                                                         const int*  attrib_list);
+      glXCreateContextAttribsARB_t aCreateCtxProc = (glXCreateContextAttribsARB_t)glXGetProcAddress(
+        (const GLubyte*)"glXCreateContextAttribsARB");
+      if (!theCaps->contextCompatible)
+      {
+        int aCoreCtxAttribs[] = {GLX_CONTEXT_MAJOR_VERSION_ARB,
+                                 3,
+                                 GLX_CONTEXT_MINOR_VERSION_ARB,
+                                 2,
+                                 GLX_CONTEXT_PROFILE_MASK_ARB,
+                                 GLX_CONTEXT_CORE_PROFILE_BIT_ARB,
+                                 GLX_CONTEXT_FLAGS_ARB,
+                                 theCaps->contextDebug ? GLX_CONTEXT_DEBUG_BIT_ARB : 0,
+                                 0,
+                                 0};
+
+        // try to create the core profile of highest OpenGL version supported by OCCT
+        for (int aLowVer4 = 6; aLowVer4 >= 0 && aGContext == NULL; --aLowVer4)
+        {
+          aCoreCtxAttribs[1] = 4;
+          aCoreCtxAttribs[3] = aLowVer4;
+          aGContext          = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCoreCtxAttribs);
+        }
+        for (int aLowVer3 = 3; aLowVer3 >= 2 && aGContext == NULL; --aLowVer3)
+        {
+          aCoreCtxAttribs[1] = 3;
+          aCoreCtxAttribs[3] = aLowVer3;
+          aGContext          = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCoreCtxAttribs);
+        }
+        isCoreProfile = aGContext != NULL;
+      }
+
+      if (aGContext == NULL)
+      {
+        int aCtxAttribs[] = {GLX_CONTEXT_FLAGS_ARB,
+                             theCaps->contextDebug ? GLX_CONTEXT_DEBUG_BIT_ARB : 0,
                              0,
-                             GL_DEBUG_SEVERITY_MEDIUM,
-                             aMsg);
-  }
+                             0};
+        isCoreProfile     = Standard_False;
+        aGContext         = aCreateCtxProc(aDisp, anFBConfig, aSlaveCtx, True, aCtxAttribs);
 
-  myGlContext->Init((Aspect_Drawable)aWindow,
-                    (Aspect_Display)aDisp,
-                    (Aspect_RenderingContext)aGContext,
-                    isCoreProfile);
-  #else
+        if (aGContext != NULL && !theCaps->contextCompatible)
+        {
+          TCollection_ExtendedString aMsg(
+            "OpenGl_Window::CreateWindow: core profile creation failed.");
+          myGlContext->PushMessage(GL_DEBUG_SOURCE_APPLICATION,
+                                   GL_DEBUG_TYPE_PORTABILITY,
+                                   0,
+                                   GL_DEBUG_SEVERITY_LOW,
+                                   aMsg);
+        }
+      }
+      XSetErrorHandler(anOldHandler);
+    }
+
+    if (myOwnGContext && aGContext == NULL)
+    {
+      aGContext = glXCreateContext(aDisp, aVis.get(), aSlaveCtx, GL_TRUE);
+      if (aGContext == NULL)
+      {
+        throw Aspect_GraphicDeviceDefinitionError(
+          "OpenGl_Window::CreateWindow: glXCreateContext failed.");
+      }
+    }
+
+    // check Visual for OpenGl context's parameters compatibility
+    TCollection_ExtendedString aList;
+    int                        isDoubleBuffer = 0, isRGBA = 0, isStereo = 0;
+    int                        aDepthSize = 0, aStencilSize = 0;
+    glXGetConfig(aDisp, aVis.get(), GLX_RGBA, &isRGBA);
+    glXGetConfig(aDisp, aVis.get(), GLX_DOUBLEBUFFER, &isDoubleBuffer);
+    glXGetConfig(aDisp, aVis.get(), GLX_STEREO, &isStereo);
+    glXGetConfig(aDisp, aVis.get(), GLX_DEPTH_SIZE, &aDepthSize);
+    glXGetConfig(aDisp, aVis.get(), GLX_STENCIL_SIZE, &aStencilSize);
+    if (aDepthSize < 1)
+      addMsgToList(aList, "no depth buffer");
+    if (aStencilSize < 1)
+      addMsgToList(aList, "no stencil buffer");
+    if (isRGBA == 0)
+      addMsgToList(aList, "no RGBA color buffer");
+    if (isDoubleBuffer == 0)
+      addMsgToList(aList, "no Double Buffer");
+    if (theCaps->contextStereo && isStereo == 0)
+    {
+      addMsgToList(aList, "no Quad Buffer");
+    }
+    else if (!theCaps->contextStereo && isStereo == 1)
+    {
+      addMsgToList(aList, "extra Quad Buffer");
+    }
+    if (!aList.IsEmpty())
+    {
+      TCollection_ExtendedString aMsg =
+        TCollection_ExtendedString("OpenGl_Window::CreateWindow: window Visual is incomplete: ")
+        + aList;
+      myGlContext->PushMessage(GL_DEBUG_SOURCE_APPLICATION,
+                               GL_DEBUG_TYPE_OTHER,
+                               0,
+                               GL_DEBUG_SEVERITY_MEDIUM,
+                               aMsg);
+    }
+
+    myGlContext->Init((Aspect_Drawable)aWindow,
+                      (Aspect_Display)aDisp,
+                      (Aspect_RenderingContext)aGContext,
+                      isCoreProfile);
+  }
+  #endif
+  #ifdef COMPILE_NOOP
   // not implemented
   (void)isCoreProfile;
   #endif
@@ -772,12 +784,13 @@ OpenGl_Window::~OpenGl_Window()
   // release "GL" context if it is owned by window
   // Mesa implementation can fail to destroy GL context if it set for current thread.
   // It should be safer to unset thread GL context before its destruction.
-  #if defined(HAVE_EGL)
-  if ((EGLSurface)myGlContext->myWindow != EGL_NO_SURFACE)
+  #ifdef COMPILE_EGL)
+  if (!RUN_UNIX_X11 && (EGLSurface)myGlContext->myWindow != EGL_NO_SURFACE)
   {
     eglDestroySurface((EGLDisplay)myGlContext->myDisplay, (EGLSurface)myGlContext->myWindow);
   }
-  #elif defined(_WIN32)
+  #endif
+  #ifdef COMPILE_WIN32
   HWND  aWindow         = (HWND)myGlContext->myWindow;
   HDC   aWindowDC       = (HDC)myGlContext->myDisplay;
   HGLRC aWindowGContext = (HGLRC)myGlContext->myGContext;
@@ -794,24 +807,29 @@ OpenGl_Window::~OpenGl_Window()
     wglDeleteContext(aWindowGContext);
   }
   ReleaseDC(aWindow, aWindowDC);
-  #elif defined(HAVE_XLIB)
-  Display*   aDisplay        = (Display*)myGlContext->myDisplay;
-  GLXContext aWindowGContext = (GLXContext)myGlContext->myGContext;
-  GLXContext aThreadGContext = glXGetCurrentContext();
-  myGlContext.Nullify();
-
-  if (aDisplay != NULL)
+  #endif
+  #ifdef COMPILE_XLIB
+  if (!RUN_WAYLAND)
   {
-    if (aThreadGContext == aWindowGContext)
-    {
-      glXMakeCurrent(aDisplay, None, NULL);
-    }
+    Display*   aDisplay        = (Display*)myGlContext->myDisplay;
+    GLXContext aWindowGContext = (GLXContext)myGlContext->myGContext;
+    GLXContext aThreadGContext = glXGetCurrentContext();
+    myGlContext.Nullify();
 
-    // FSXXX sync necessary if non-direct rendering
-    glXWaitGL();
-    glXDestroyContext(aDisplay, aWindowGContext);
+    if (aDisplay != NULL)
+    {
+      if (aThreadGContext == aWindowGContext)
+      {
+        glXMakeCurrent(aDisplay, None, NULL);
+      }
+
+      // FSXXX sync necessary if non-direct rendering
+      glXWaitGL();
+      glXDestroyContext(aDisplay, aWindowGContext);
+    }
   }
-  #else
+  #endif
+  #ifdef COMPILE_NOOP
     // not implemented
   #endif
 }
@@ -855,43 +873,46 @@ void OpenGl_Window::init()
     return;
   }
 
-  #if defined(HAVE_EGL)
-  if ((EGLSurface)myGlContext->myWindow == EGL_NO_SURFACE)
+  #if defined(HAVE_EGL) || WAYLAND_COMPAT
+  if (!RUN_UNIX_X11)
   {
-    // define an offscreen default FBO to avoid rendering into EGL_NO_SURFACE;
-    // note that this code is currently never called, since eglCreatePbufferSurface() is used
-    // instead as more robust solution for offscreen rendering on bugged OpenGL ES drivers
-    Handle(OpenGl_FrameBuffer) aDefFbo =
-      myGlContext->SetDefaultFrameBuffer(Handle(OpenGl_FrameBuffer)());
-    if (!aDefFbo.IsNull())
+    if ((EGLSurface)myGlContext->myWindow == EGL_NO_SURFACE)
     {
-      aDefFbo->Release(myGlContext.operator->());
-    }
-    else
-    {
-      aDefFbo = new OpenGl_FrameBuffer();
-    }
+      // define an offscreen default FBO to avoid rendering into EGL_NO_SURFACE;
+      // note that this code is currently never called, since eglCreatePbufferSurface() is used
+      // instead as more robust solution for offscreen rendering on bugged OpenGL ES drivers
+      Handle(OpenGl_FrameBuffer) aDefFbo =
+        myGlContext->SetDefaultFrameBuffer(Handle(OpenGl_FrameBuffer)());
+      if (!aDefFbo.IsNull())
+      {
+        aDefFbo->Release(myGlContext.operator->());
+      }
+      else
+      {
+        aDefFbo = new OpenGl_FrameBuffer();
+      }
 
-    OpenGl_ColorFormats aColorFormats;
-    aColorFormats.Append(GL_RGBA8);
-    if (!aDefFbo->InitRenderBuffer(myGlContext, mySize, aColorFormats, GL_DEPTH24_STENCIL8))
+      OpenGl_ColorFormats aColorFormats;
+      aColorFormats.Append(GL_RGBA8);
+      if (!aDefFbo->InitRenderBuffer(myGlContext, mySize, aColorFormats, GL_DEPTH24_STENCIL8))
+      {
+        TCollection_AsciiString aMsg("OpenGl_Window::CreateWindow: default FBO creation failed");
+        throw Aspect_GraphicDeviceDefinitionError(aMsg.ToCString());
+      }
+      myGlContext->SetDefaultFrameBuffer(aDefFbo);
+      aDefFbo->BindBuffer(myGlContext);
+    }
+    else if (!myPlatformWindow->IsVirtual() && mySizeWindow == myPlatformWindow)
     {
-      TCollection_AsciiString aMsg("OpenGl_Window::CreateWindow: default FBO creation failed");
-      throw Aspect_GraphicDeviceDefinitionError(aMsg.ToCString());
+      eglQuerySurface((EGLDisplay)myGlContext->myDisplay,
+                      (EGLSurface)myGlContext->myWindow,
+                      EGL_WIDTH,
+                      &mySize.x());
+      eglQuerySurface((EGLDisplay)myGlContext->myDisplay,
+                      (EGLSurface)myGlContext->myWindow,
+                      EGL_HEIGHT,
+                      &mySize.y());
     }
-    myGlContext->SetDefaultFrameBuffer(aDefFbo);
-    aDefFbo->BindBuffer(myGlContext);
-  }
-  else if (!myPlatformWindow->IsVirtual() && mySizeWindow == myPlatformWindow)
-  {
-    eglQuerySurface((EGLDisplay)myGlContext->myDisplay,
-                    (EGLSurface)myGlContext->myWindow,
-                    EGL_WIDTH,
-                    &mySize.x());
-    eglQuerySurface((EGLDisplay)myGlContext->myDisplay,
-                    (EGLSurface)myGlContext->myWindow,
-                    EGL_HEIGHT,
-                    &mySize.y());
   }
   #else
     //
diff --git c/src/Standard/FILES w/src/Standard/FILES
index c5d5af44cb..c6a38d3cad 100755
--- c/src/Standard/FILES
+++ w/src/Standard/FILES
@@ -44,6 +44,7 @@ Standard_JmpBuf.hxx
 Standard_LicenseError.hxx
 Standard_LicenseNotFound.hxx
 Standard_Macro.hxx
+Standard_Macro_Runtime.hxx
 Standard_MemoryUtils.hxx
 Standard_MMgrOpt.cxx
 Standard_MMgrOpt.hxx
diff --git c/src/Standard/Standard_Macro_Runtime.hxx w/src/Standard/Standard_Macro_Runtime.hxx
new file mode 100644
index 0000000000..830e939f6a
--- /dev/null
+++ w/src/Standard/Standard_Macro_Runtime.hxx
@@ -0,0 +1,75 @@
+
+// Copyright (c) 1998-1999 Matra Datavision
+// Copyright (c) 1999-2014 OPEN CASCADE SAS
+//
+// This file is part of Open CASCADE Technology software library.
+//
+// This library is free software; you can redistribute it and/or modify it under
+// the terms of the GNU Lesser General Public License version 2.1 as published
+// by the Free Software Foundation, with special exception defined in the file
+// OCCT_LGPL_EXCEPTION.txt. Consult the file LICENSE_LGPL_21.txt included in OCCT
+// distribution for complete text of the license and disclaimer of any warranty.
+//
+// Alternatively, this file may be used under the terms of Open CASCADE
+// commercial license or contractual agreement.
+
+//! @file
+//! This file provides macros to determine system runtime compatibility
+//! capabilities. It's primary purpose is to compile both X11 and Wayland
+//! paths for Unix-like systems thus allowing the consumer to select the
+//! appropriate runtime path.
+
+#ifndef _Standard_Macro_Runtime_HeaderFile
+#define _Standard_Macro_Runtime_HeaderFile
+
+//! @def CHECK_ENV_VAL
+//! Compares the value of an environment variable
+//! @returns bool
+#define CHECK_ENV_VAL(key, val)                                                                    \
+  ([&]() -> bool {                                                                                 \
+    const char* result = std::getenv(key);                                                         \
+    return result != nullptr && std::strcmp(result, val) == 0;                                     \
+  }())
+
+//! @def RUN_UNIX_X11
+//! Determine a X11 runtime path on Wayland compatible systems
+//! @returns bool
+#define RUN_UNIX_X11                                                                               \
+  ([]() -> bool {                                                                                  \
+    if (!WAYLAND_COMPAT)                                                                           \
+      return 0;                                                                                    \
+                                                                                                   \
+    return CHECK_ENV_VAL("XDG_SESSION_TYPE", "x11") || CHECK_ENV_VAL("GDK_BACKEND", "x11")         \
+           || CHECK_ENV_VAL("QT_QPA_PLATFORM", "xcb");                                             \
+  }())
+
+//! @def RUN_WAYLAND
+//! Determine a Wayland runtime path on Wayland compatible systems
+//! @returns bool
+#define RUN_WAYLAND                                                                                \
+  ([]() -> bool {                                                                                  \
+    if (!WAYLAND_COMPAT || RUN_UNIX_X11)                                                           \
+      return 0;                                                                                    \
+                                                                                                   \
+    return CHECK_ENV_VAL("XDG_SESSION_TYPE", "wayland") || CHECK_ENV_VAL("GDK_BACKEND", "wayland") \
+           || CHECK_ENV_VAL("QT_QPA_PLATFORM", "wayland");                                         \
+  }())
+
+// Compiler definitions to negate elif macro patterns downstream.
+// This allows for compiling both X11 and Wayland paths.
+#if WAYLAND_COMPAT
+  #define COMPILE_EGL
+  #define COMPILE_XLIB
+#elif defined(HAVE_EGL)
+  #define COMPILE_EGL
+#elif defined(_WIN32)
+  #define COMPILE_WIN32
+#elif defined(HAVE_XLIB)
+  #define COMPILE_XLIB
+#elif defined(__APPLE__)
+  #define COMPILE_APPLE
+#else
+  #define COMPILE_NOOP
+#endif
+
+#endif // _Standard_Macro_Runtime_HeaderFile
